### ISSUE TITLE
TestClusterResourceSetReconciler test is flaky

### ISSUE THREAD

**Issue by Sunnatillo** [kubernetes-sigs/cluster-api#10854]
### Which jobs are flaking?

capi-test-main

### Which tests are flaking?

TestClusterResourceSetReconciler

### Since when has it been flaking?

Most likely after merging this PR: https://github.com/kubernetes-sigs/cluster-api/pull/10656

### Testgrid link

edited: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-test-mink8s-main/1810533764078505984

### Label(s) to be applied

/kind flake
/area ci

**Comment by sbueringer** [comment]
cc @fabriziopandini Should be the same as the one I mentioned last time

**Comment by sbueringer** [comment]
@Sunnatillo I think the link is pointing to the wrong job

**Comment by Sunnatillo** [comment]
Fixed now. 
Also another occurence: 
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-test-main/1807033132835147776

**Comment by Sunnatillo** [comment]
> @Sunnatillo I think the link is pointing to the wrong job

Fixed now.
Also another occurence:
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-test-main/1807033132835147776

**Comment by fabriziopandini** [comment]
@jimmidyson PTAL
I will also take a look because I recently made changes on ownerReference management on the same area of the code base

It is important to nail this down before release

**Comment by jimmidyson** [comment]
I can't see results before 30th June to check for flakiness before https://github.com/kubernetes-sigs/cluster-api/pull/10756 was merged on 28th June. Is that data available? The first flake appeared in f57b8c8ef which was merged after that PR too, but like I said I can't tell if the build was stable before that PR or not 😞

**Comment by chrischdi** [comment]
This page here should allow you to go further back:

https://prow.k8s.io/job-history/gs/kubernetes-jenkins/logs/periodic-cluster-api-test-mink8s-main?buildId=

As alternative, you could try to filter at [k8s-triage](https://storage.googleapis.com/k8s-triage/index.html?job=.*-cluster-api-.*&xjob=.*-provider-.*)

**Comment by jimmidyson** [comment]
Thanks @chrischdi!

I've gone back to the time of merge of #10656 and I can only see failures for this test after #10756 was merged so I'm pretty sure that introduced the flakiness. I'll take a closer look if I can find time to try to help figure out what's going on.

**Comment by Sunnatillo** [comment]
Increasing timeout helps to solve the issues.

**Comment by fabriziopandini** [comment]
The fact that a ClusterResourceSet binding takes so long to reach a stable state isn't ideal.

The issue is on the fact that we are re-queuing in case of API conflicts, and then next reconciliations are influenced by exponential backoff delay quicky growing (+ the other side of the coin, that many reconcilation are happening in a very short sequence at the beginning of the backoff sequence).

TL;DR exponential backoff should be used to handle errors, not to handle how controllers are reaching a stable state

I have submitted https://github.com/kubernetes-sigs/cluster-api/pull/10869 to get rid of the exponential backoff and all of the noise that API conflict were adding to the logs + documented the problem

But also in this case this is a mitigation.

**Comment by sbueringer** [comment]
Let's see if the test is stable now and then close the issue in 1-2 days if the test is stable

(xref: https://github.com/kubernetes-sigs/cluster-api/pull/10868#issuecomment-2233489680)

**Comment by sbueringer** [comment]
Link to check for new occurences https://storage.googleapis.com/k8s-triage/index.html?text=Should_handle_applying_multiple_ClusterResourceSets_concurrently_to_the_same_cluster&job=.*cluster-api.*main&xjob=.*-e2e-.*%7C.*-provider-.*

**Comment by Sunnatillo** [comment]
> Link to check for new occurences https://storage.googleapis.com/k8s-triage/index.html?text=Should_handle_applying_multiple_ClusterResourceSets_concurrently_to_the_same_cluster&job=.*cluster-api.*main&xjob=.*-e2e-.*%7C.*-provider-.*

Not occurred today. It was occurred often before the fix. We can close this issue and the PR.

**Comment by Sunnatillo** [comment]
I did the test with 200 count, all passed. I would say we are safe to close this issue.

**Comment by sbueringer** [comment]
Great! Thx for testing

/close
